chain/chain/src/validate.rs

```diff
@@ -124,27 +123,31 @@ pub fn validate_chunk_with_chunk_extra(
     prev_chunk_header: &ShardChunkHeader,
     chunk_header: &ShardChunkHeader,
 ) -> Result<(), Error> {
-    if prev_chunk_extra.state_root != chunk_header.prev_state_root() {
+    if *prev_chunk_extra.state_root() != chunk_header.prev_state_root() {
         return Err(ErrorKind::InvalidStateRoot.into());
     }
 
-    if prev_chunk_extra.outcome_root != chunk_header.outcome_root() {
+    if *prev_chunk_extra.outcome_root() != chunk_header.outcome_root() {
         return Err(ErrorKind::InvalidOutcomesProof.into());
     }
 
-    if prev_chunk_extra.validator_proposals != chunk_header.validator_proposals() {
+    let chunk_extra_proposals = prev_chunk_extra.validator_proposals();
+    let chunk_header_proposals = chunk_header.validator_proposals();
+    if chunk_header_proposals.len() != chunk_extra_proposals.len()
+        || !chunk_extra_proposals.zip(chunk_header_proposals).all(|(a, b)| a == b)
```

Comment:
impl PartialEq/Eq on ValidatorStakeIter and simply compare `!=` here?

Reply:
I don't think implementing `PartialEq` makes sense because the `eq` method signature uses `&self`, whereas to access elements in the iterator (via `all` or `next`) I need `&mut self`. This is also the only place where we compare two of these objects, so it's not saving much to implement the logic elsewhere.

Reply:
You might make use of https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.eq though I think.

Reply:
Ah, yes, that is nicer! Done in https://github.com/near/nearcore/pull/4115/commits/321c4cb9569ade9daf2dabf3ba52eb465c9760f1